New User Experience (NUX)

noisysocks opened the pull request:
### 🕺 What this is

Closes #3670. Blocked by https://github.com/WordPress/gutenberg/pull/7038.

This implements the NUX flow outlined by @karmatosed in https://github.com/WordPress/gutenberg/issues/3670#issuecomment-384595696:

When a new user opens Gutenberg, they're presented with a guide that points out some useful page elements:

![nux](https://user-images.githubusercontent.com/612155/40633279-0d4e915c-6332-11e8-80e2-39f10f2c838e.gif)

If they dismiss the guide, it won't ever show again:

![nux-dismiss](https://user-images.githubusercontent.com/612155/40633389-ccc2127a-6332-11e8-991b-004f5f0da523.gif)

### 🤔 How it works

- A `DotTip` component renders a tip bubble independently of anything else
  - This uses `Popover`, which now supports having its `yAxis` set to `middle`
- The `triggerGuide()` action lets you associate a series of tips into a single guide
- The `isTipVisible()` and `getAssociatedGuide()` selectors return information about which step in a guide the user is in
- The `dismissTip()` and `disableTips()` actions control dismissing a single tip (and stepping through the guide) and disabling tips entirely
- All of this functionality is contained within a new `nux` module with the intention that these components should be made available in other areas of the WP Admin

### 📋 Testing

There's unit tests included. Manual testing can be done by:

1. Create a new post
2. View and step through the tips
3. Refresh the page—the current step should be maintained 
3. Dismiss the tips
4. Refresh the page—the tips should no longer appear

### ✍️ Copy

I'm still waiting on the final copy for each tip. There was some discussion about this in https://github.com/WordPress/gutenberg/issues/3670.

noisysocks commented:
cc. @karmatosed

Soean commented on editor/components/new-user-tip/index.js:
```diff
@@ -0,0 +1,107 @@
+	},
+	preview: {
+		step: 3,
+		text: __( 'Click ‘Preview’ to load a preview of this page, so you can make sure you’re happy with your blocks. ' ),
```
Whitespace at the end of the sentence.

noisysocks replied:
Thanks, removed in 4b9ff4d80.

youknowriad commented:
Just a quick comment :) I'd love if the nux is built as a separate module to be reusable in all WP and not only the editor context. See https://github.com/WordPress/gutenberg/pull/4173 for prior art. I started this module and put it on hold because of priorities :)

noisysocks commented:
> Just a quick comment :) I'd love if the nux is built as a separate module to be reusable in all WP and not only the editor context. See #4173 for prior art. I started this module and put it on hold because of priorities :)

Oh, nice. I'll move this into a seperate `nux` module.

noisysocks commented:
This is in a reviewable state now—I've updated the PR description 🐶🌈✨

karmatosed commented:
This is coming on well. I have some initial bits of feedback and then I think we can really move into polish.

- I feel the width could be increased on the tips we need to of course adapt to mobile, but the small size is making an issue with visuals, this isn't a tooltip and shouldn't have those size restrictions.

![2018-05-22 at 17 58](https://user-images.githubusercontent.com/253067/40374796-17f8fc66-5dea-11e8-8d22-90975fbdaa51.png)

- Can we point out the '+' in the post content over being all at the top bar like the mocks?

![2018-05-22 at 18 01](https://user-images.githubusercontent.com/253067/40374832-3d0cd96e-5dea-11e8-8977-ed7e242cf5f3.png)

- Can we not have it ever go over the toolbar?

![2018-05-22 at 17 59](https://user-images.githubusercontent.com/253067/40374891-60f9a942-5dea-11e8-9ad3-748c5fc1815c.png)

A couple of things, I would like it to be a link not a button, the button to move to next tip. That was a link on purpose to make it more conversational than a required UI. I also would say let's iterate on what it says 'got it' feels a little too casual. @michelleweber what would you suggest for the words to move to next tip?

youknowriad commented on edit-post/components/header/header-toolbar/index.js:
```diff
@@ -30,7 +31,11 @@ function HeaderToolbar( { hasFixedToolbar, isLargeViewport } ) {
 		>
-			<Inserter position="bottom right" />
+			<Inserter position="bottom right">
+				<GuideTip guideID="core/editor" step={ 1 }>
```
Do we really want to step prop? I'm thinking these tips should be autonomous because I can see that some may show up where others not depending on different criterias:

 - The blocks count
 - Whether it's the first time we're using gutenberg
 - Whether it's the "n" time we use gutenberg (more experienced tips)
 - Whether we open a classic block's post (tip to indicate that it's a classic block that can be converted)
 - ...

My idea is that tips have only ids, I can even see several tips showing at the same time some time.
I think we probably need guides too but I wonder if it's better to use props for this or another API like:

```
const guide = [ tipId1, tipId2];
wp.dispatch( 'core/nux' ).triggerGuide( guide ):
```

I'm just thinking outloud here, nothing very well thought but I admit I find a the step prop problematic.

noisysocks replied:
Agree that more autonomous tips would be more future-proof—I'll have a play with something like `triggerGuide()`.

noisysocks commented:
Thanks @karmatosed! In 66ba2ac18 I:

- Increased the width of tips by 100px
- Made tips not appear over the toolbar
- Styled the next button as a link

> Can we point out the '+' in the post content over being all at the top bar like the mocks?

It's difficult because that button doesn't appear until you hover your mouse over the block placeholder. I'll see what I can do, though.

> To be super-clear, I'd probably go with "Next" or "What's next?" on a button. If you want it to be a text link instead, either "What's next?" or "Great! What's next?" or something like "Tell me more!"

My issue with copy that contains the word 'next' is that, when you're on the last tip, there isn't a 'next tip' to go to. I've changed the copy to 'Tell me more!' for now but will work on making it so that we can have different copy for the button when it's the last tip.

>

noisysocks commented on edit-post/components/layout/index.js:
```diff
@@ -94,6 +95,7 @@ function Layout( {
 			}
 			<Popover.Slot />
 			<PluginArea />
+			<GuideTip.Slot />
```
I moved away from `Popover` as there’s some differences in appearance and behaviour. We probably will want to extract common functionality into a shared component e.g. `PopoverBase`. I’m wary of introducing this abstraction while we’re experimenting with the NUX UI, though.

youknowriad replied:
I just made some adjustments to the Popover component and I think it's great in resolving all the positioning/size issues. I'd love to know what are those difference to see if we can bake them into the Popover component. From the screenshots it looks very similar to the popover component. I believe what's missing is the support of the "center" position in the yAxis. We only support it in the `xAxis`.

noisysocks replied:
I’ll check out your adjustments. When I last looked, the differences were:

- Blue dot instead of an arrow
- Should appear to the left or right of the anchor point instead of above or below
- Clicking outside shouldn’t dismiss

youknowriad replied:
> Blue dot instead of an arrow

The blue dot should be the anchor (outside the popover)

> Should appear to the left or right of the anchor point instead of above or below

Yes, that's what's missing the "center" support for the yAxis. I believe we can add it.

> Clicking outside shouldn’t dismiss

This should work, maybe a bug somewhere

> Must appear above other popovers

Popovers take classNames, I believe we can adjust the z-index for nux popovers

youknowriad replied:
I'm adding a `noArrow` prop to this one that could also help https://github.com/WordPress/gutenberg/pull/6911#issuecomment-391325040

karmatosed commented:
@noisysocks thanks for the updates. I do have one adjustment that will effect all I see, can we have more space between the pulsing dot and area it's indicating? Here are some adjusted mocks - the last one in each group is the one I'd like to change to.

![artboard1](https://user-images.githubusercontent.com/253067/40515834-a577c340-5fae-11e8-8eeb-c4816cdaa5e0.png)

![artboard 2](https://user-images.githubusercontent.com/253067/40515838-a99a2e0e-5fae-11e8-94a7-0b2a443ec549.png)

![artboard 3](https://user-images.githubusercontent.com/253067/40515839-a9c669ec-5fae-11e8-8e0a-3b6f75070949.png)

youknowriad commented on components/popover/index.js:
```diff
@@ -213,6 +213,11 @@ class Popover extends Component {
 		} = this.state;
 
+		let { noArrow = false } = this.props;
+		if ( xAxis === 'center' && yAxis === 'middle' ) {
```
Not certain about this? in which case will have both centered?

noisysocks replied:
Perhaps a `console.warn` would be better? I just want to explicitly declare somewhere that `position="middle center" noArrow={ false }` is something that the Popover component does **not** support.

youknowriad commented on components/popover/utils.js:
```diff
@@ -91,7 +101,14 @@ export function computePopoverYAxisPosition( anchorRect, contentSize, yAxis ) {
+		popoverTop = topAlignment.popoverTop;
+	} else {
+		popoverTop = bottomAlignment.popoverTop;
+	}
```
Thanks for the updates 👍 Can we add some unit tests for the "middle" alignment?

noisysocks committed:
Fix appearance of DotTip on mobile

Force DotTip content to appear underneath the indicator dot when on
mobile viewports. This requires overriding a lot of .component-popover
styles which is quite gross.

noisysocks committed:
Remove unnecessary <Fragment>

noisysocks committed:
Position NUX dots right next to the edge of the button

Adjust popover's layout algorithm so that if yAxis === 'middle', it will
set popoverLeft such that the popover points to the edge of the anchor
node. This aligns with what e.g. position="bottom center" does.

noisysocks commented:
OK, I think that this is in a good place now. I've updated the PR description.

---

@karmatosed: I struggled a little with getting the dots to appear exactly how you want. To keep the code simple, we should try to position the dot in the same way for every control. I think it's best that we pick a single number of pixels between the centre of the dot and the left/right edge of the button. I went with 4px which, since the dot is 8px in diameter, means that the dot rests flush against the edge of the button.

To illustrate what I mean:

<img width="639" alt="margin 1" src="https://user-images.githubusercontent.com/612155/40633632-3833dbe6-6334-11e8-8e12-8b0bc9db45ac.png">

<img width="500" alt="margin 2" src="https://user-images.githubusercontent.com/612155/40633640-4c3097f6-6334-11e8-8438-630368ecaa83.png">

This is how it looks:

![dot-inserter](https://user-images.githubusercontent.com/612155/40633675-8616dce6-6334-11e8-937d-0ba74d0c15e0.gif)

![dot-settings](https://user-images.githubusercontent.com/612155/40633695-a25fa78e-6334-11e8-9fd5-b6ff6bf1ded8.gif)

![dot-preview](https://user-images.githubusercontent.com/612155/40633710-af95418e-6334-11e8-8026-a315fa1ebba3.gif)

Let me know if this is OK or if you'd like to change the number of pixels or the approach I'm taking.

karmatosed commented:
Here is a video for those who want to follow along and review:

![tips](https://user-images.githubusercontent.com/253067/40670165-993079f0-6360-11e8-806a-10a5a84e1be4.gif)

karmatosed commented:
Here is what it looks like on non-desktop screens:

![tipsmobile](https://user-images.githubusercontent.com/253067/40670339-fd06f13e-6360-11e8-97e8-eede7615a4b1.gif)

noisysocks committed:
Increase NUX tip padding

noisysocks committed:
Add margin to sides of NUX tip on mobile

noisysocks commented:
Thanks for the feedback @melchoyce!

> I want the tip to stand out more — whether that's dimming the background behind it, except for whatever element is being called out, or even adding a splash of color to the tip, like a solid border on one side.

Any ideas or preference, @karmatosed? I'm scared that a dimmed background would trigger the PTSD I have from visiting websites that have pop-over ads 😨 

> The left and right padding inside the tip feels kind of tight.

👌 I've bumped this up in 6d85e3b80.

> The dot doesn't stand out when focusing on the publish button, since they're both blue. (Might not be fixable, just the nature of it.)

> “Tell me more” makes me think clicking that link will open a new window that goes into more detail on whatever is being highlighted. Maybe "continue" instead?

Those screenshots @karmatosed posted are a little out of date.

The button copy was changed to 'Show next tip' for the first three tips and then 'Got it' for the last tip.

The blue dot issue might be a little better now that the blue dot is positioned more to the left of the _Publish_ button:

<img width="658" alt="nux-publish" src="https://user-images.githubusercontent.com/612155/40692232-ecff31f0-63f3-11e8-9c92-489bdd123713.png">

> On mobile, I wish there was a little bit of space around the edge of the tip so it's not flush against the sides of the screen.

👌 added some margins in cb9db5595.

noisysocks commented:
Added this to the 3.0 milestone since that's our (ambitious) goal. 

@youknowriad—how does this feel, code-wise?

@karmatosed @michelleweber—could you please send me the final copy that we want to use for the tips?

@afercia—would you mind taking a quick look at this from an a11y perspective? I followed the considerations noted in https://github.com/WordPress/gutenberg/issues/3670#issuecomment-384600658.

youknowriad commented on nux/README.md:
````diff
@@ -0,0 +1,86 @@
+A tip cannot be added to more than one guide.
+
+```jsx
+domReady(() => {
````
Should we drop `domReady` here? I don't see the relation ship with guides. I tend to think guides should be triggered in `componentDidMount` methods instead or init functions like you did.

noisysocks replied:
👌 9acf4167d

youknowriad commented on nux/components/dot-tip/index.js:
```diff
@@ -0,0 +1,90 @@
+		if ( this.props.isVisible ) {
+			// Fix the tip not appearing next to the inserter toggle by forcing Popover
+			// to recalculate its size and position on the next frame
+			defer( () => {
```
Why is this necessary? do you think this is a Popover bug instead and should be fixed there?

noisysocks replied:
The first tip that appears on page load is mounted before layout has finished and this messes up the position calculation. I've moved the fix into Popover where it looks a littles less hacky in 6d8173b2d

noisysocks replied:
OK, I dug into this.

The problem is that, during the initial page load, the popover position is calculated while `PostTitle` is hidden because `PostTypeSupportCheck` is waiting for the `getPostType` resolver to complete:

<img width="1112" alt="screen shot 2018-05-31 at 15 25 08" src="https://user-images.githubusercontent.com/612155/40763216-6e1f2828-64e7-11e8-89a6-81ff2363b67f.png">

When the resolver eventually finishes, `PostTitle` appears and the popover is now in the wrong position:

<img width="1111" alt="screen shot 2018-05-31 at 15 30 27" src="https://user-images.githubusercontent.com/612155/40763256-a3ad744a-64e7-11e8-8646-49bbdb6cd75b.png">

To fix this properly, `Popover` needs to recalculate its position whenever its anchor moves (e.g. because of a DOM change).

I'm not sure how best to do that. Two ideas:

1. We could use `MutationObserver` to watch for changes to the DOM. This would be quite expensive though since we'd have to trigger a recalculation whenever **any** of the anchor's parents changes. We could mitigate the performance penalty by throttling the recalculations and/or making it opt-in (e.g. so that only the first `DotTip` does this). 
2. We could recalculate the position e.g. in a `setInterval`. Bad idea—sorry, that's all I've got.

Thoughts? Alternative ideas?

youknowriad replied:
Yes, that's not an easy problem to solve. I wonder why `getPostType` is not resolving immediately, I thought we were preloading all the required data on initial load.

cc @aduth

noisysocks replied:
Let's leave the workaround as is for now. I've updated the inline comment to better explain the problem and created a issue to follow-up on here: https://github.com/WordPress/gutenberg/issues/7071

youknowriad commented on nux/store/selectors.js:
```diff
@@ -0,0 +1,57 @@
+		return false;
+	}
+
+	return true;
```
I'm thinking that if the tip is not associated with a guide, it should be hidden by default unless we click on the dot to open it. And it should be dismissed entirely (with the dot) once we click the "close" button. What do you think?

youknowriad commented on lib/client-assets.php:
```diff
@@ -402,7 +410,7 @@ function gutenberg_register_scripts_and_styles() {
 		'wp-editor',
 		gutenberg_url( 'build/editor/style.css' ),
-		array( 'wp-components', 'wp-editor-font' ),
+		array( 'wp-components', 'wp-editor-font', 'wp-nux' ),
```
Should we do the same for 'wp-edit-post'?

noisysocks replied:
👌 0499f5c82

youknowriad commented on edit-post/index.js:
```diff
@@ -68,6 +69,13 @@ export function initializeEditor( id, post, settings ) {
+		'core/editor.settings',
+		'core/editor.preview',
+		'core/editor.publish',
+	] );
```
I noticed that when dismissing a notice in a guide, and the next tip is not present in the DOM, no tip is being shown. Shouldn't we just skip to the next available tip?

noisysocks replied:
I don't think so. A tip in the guide might rely on the user having read a previous tip, e.g. _Now that you've learned about X, this is what Y does!_

youknowriad reviewed (commented):
Left some comments, but code wise, this is solid. 👏

afercia commented:
@noisysocks thanks for the ping and thanks for your work, it looks very nice. There are a few a11y things to address but first, some non-a11y things:
- I get a `Warning: validateDOMNesting(...): <button> cannot appear as a descendant of <button>` but at a first look couldn't find nested buttons; I guess there's something to check though
- I get a warning `React does not recognize the `noArrow` prop on a DOM element` so I guess this prop gets passed to some DOM element
- the Close button tooltip has a minor z-index issue (its "arrow" gets hidden by the tip box):

<img width="338" alt="tips close tooltip x-index" src="https://user-images.githubusercontent.com/1682452/40713870-572a0732-6401-11e8-95e3-1a0d88127a4d.png">

- I'm not sure the "Close" button should disable the tips; it looks like a Close button, so maybe it should just close the modal. I'd consider a separate UI control to disable the tips

a11y:
- as noted by @jasmussen I'd also think the tips box should be "modal", meaning interaction with the rest of the UI shouldn't be allowed, also keyboard navigation should be constrained within the tips box and assistive technologies should see the box as the only available content
- in other words, it should really be an ARIA modal dialog; worth noting a modal component is in the works in #6261 which provides also a HoC to constrain tabbing (also split out in a separate PR, see #6987)
- this brings us back to the ongoing conversation on modals: Gutenberg needs a component that is a fully accessible modal; whether it's made via the Popover or a separate specific Modal component, I guess it should be re-usable also here for the tips
- initial focus: when I click on New Post, the initial tip appears but focus is at the root of the document; this is not so ideal for keyboard and assistive technology users. In the same way the tips are visually prominent and draw attention, they should also be the first content users perceive so initial focus should go to the modal
- I have some doubts about modals that self-initiate 🙂 They're not so different from ads that overlay a web page. I remember a conversation with @karmatosed, about the difference between modals that are intentionally triggered by users and modals that auto-appear on page load. Are we sure the current behavior is ideal? Wouldn't a button "Take a tour" be a better experience? This way, the tips would appear after an explicit users action.
- despite my initial recommendation, we should remove `aria-modal` because Safari 11 has a weird bug, see https://github.com/reactjs/react-modal/issues/654#issuecomment-385944392. Of course, if we end up using a standardized Modal component, this would be built-in
- `when clicking "See next" focus should go to the following modal` yes this is implemented, but I'm not sure it should go on the "See next" button, as the previous content is basically "skipped"
- Close button position: visual order should always match DOM order; currently, this button is at the end of the tip content but then positioned with CSS at the top
- labels: I'd consider to improve the wording; the modal itself has an aria-label `New user tip` which is a bit confusing, as a user I wouldn't understand why it says "New user". It should just say "Gutenberg tips" then screen readers will also announce "dialog" or "web dialog" so "Gutenberg tips... dialog" would be clearer. The close button has an aria-label "Disable guide" (see also previous point). As a used I wouldn't know what a "guide" is, as this whole thing is called "tips". Finally, I'd consider to expand a bit the next button: "See next" doesn't explicitly tell me what "next" is. "See next tip" or just "Next tip" would be a bit clearer.
- Where focus should land when closing the tip? Currently, this scenario is not managed and focus fallbacks to the document root, which is not necessarily bad if we can't find a better place

karmatosed commented:
What a lot of awesome feedback! Let me dig in :) Apologies also for my build not being up to date and throwing everyone on screenshots.

I would like to avoid dimming the background so lets not have that for now. I wouldn't be against more of a border, my concern though is that then becomes a WordPress notice. Maybe we can iterate though in next versions and just have in what we have now?

> @karmatosed @michelleweber—could you please send me the final copy that we want to use for the tips?

I thought we had the final copy?

Regarding the bubble pointer @jasmussen I would like version one to ship without. My concern over the pointer is both it becomes cumbersome across devices and it makes this the same as admin tooltips - it's a different format. I am not saying at all what we have now is perfect but it feels like a shippable v1 we can iterate with.

Button wise, I would again caution on making it a button, that to me takes away from the conversation UI aspect we have as a flow.

> Right now you can click outside the box, even when it's showing. This can lead to all sorts of weird things I'm sure. The box is not modal, in other words.

I think having clicking outside to dismiss makes sense.

> I like tips. In video games tips are usually togglable in an Options menu. Should we make an option in the More menu that is enabled by default for new users, and is disabled when you're done or dismiss it?

Yes this was in the original spec and I would like to have it included. Maybe w can make sure this gets into v1 if desired?

> But it would be nice if we could make these more like "scripted interactions" in video game parlance. For example, instead of having you page through a sequence of steps before you can get to the fun, a tip pops up the first time you use a feature.

So much yes! I see us building on this and excited where this goes.

@afercia: 

> I have some doubts about modals that self-initiate 🙂 They're not so different from ads that overlay a web page. I remember a conversation with @karmatosed, about the difference between modals that are intentionally triggered by users and modals that auto-appear on page load. Are we sure the current behavior is ideal? Wouldn't a button "Take a tour" be a better experience? This way, the tips would appear after an explicit users action.

It was decided by everyone that having on first load made sense and also avoided clashes with the toolbar up top. Is this an accessibility deal breaker?

afercia commented:
> It was decided by everyone that having on first load made sense

Hm, we'll see. This way, seems to me they're very similar to the WP core "pointers" which were argued for years in lengthy discussions and are a bit controversial, to say the least.

> Is this an accessibility deal breaker?

We've had a lot of issues with the WordPress Theme/Plugin editor "Heads up" modal, see starting from https://core.trac.wordpress.org/ticket/42110#comment:12 as screen readers don't "react" 😆very well to modals that open on page load.

karmatosed commented:
To be clear this is what was suggested before:

![2018-05-30 at 18 36](https://user-images.githubusercontent.com/253067/40737360-66a4a0f0-6438-11e8-9728-d46529eb2cea.png)

If we did add that back in then I would like input from @michelleweber on the content. What do others think who gave insights? @jasmussen, @folletto and @melchoyce for example?

folletto commented:
_Note: this discussion is massive, and I haven't read beyond the last four messages._

> To be clear this is what was suggested before:

I suggested to not do that because:

1. It would introduce a secondary design to an interaction pattern that is already suited for that (message + button + dismiss). Two UIs for the same thing seems counter intuitive in general, and more specifically end up being two different patterns to understand and learn.
2. Starting with already a tip visible is better than asking if a tip can be given. For two reasons: one, at the very minimum, the user will read a tip, and two, **gives value immediately**. 
3. Dismissing is identical (one click) in both scenarios.
4. Code wise, "Popping in" a modal, or "Popping in" a bar, can be equivalent: if the DOM placement is there, there should be no difference in terms of a11y. 

I think most of the issue really hinges on this last point, right?

It's different of course if the element as Andrea points out is "added" dynamically in a second moment. However, while I might not be seeing technical complications, I believe it's possible to have it accessible: functionally "having a bar" and "having a modal" is just adding a single line CSS: `position: absolute;`. 

I'm sure we can solve that without reverting to an entirely different UI?

—

On everything else I think Andreas outline is excellent, as usual 🌟

aduth commented on nux/components/dot-tip/index.js:
```diff
@@ -0,0 +1,90 @@
+				position="middle right"
+				noArrow
+				focusOnMount
+				role="dialog"
```
Should it not be the responsibility of `Popover` to assign this? `aria-modal` ? `aria-label` ?

noisysocks replied:
This is a little out of my comfort zone and broaches on the work that @afercia is doing with building a modal component.

Still, my thoughts:

- `Popover` could always set `role="dialog"` because "The dialog role is used to mark up an HTML based application dialog or window that separates content or UI from the rest of the web application or page." ([source](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_dialog_role)) suggests to me that all popovers are dialogs
- `Popover` could set `aria-modal` to `"true"` if there is no `onClose` prop provided
- `Popover` _should_ always have an `aria-label`, but the caller would need to provide it since there isn't a sensible default value

aduth commented on nux/components/dot-tip/index.js:
```diff
@@ -0,0 +1,90 @@
+	withDispatch( ( dispatch, { id } ) => {
+		const { dismissTip, disableTips } = dispatch( 'core/nux' );
+		return {
+			onDismiss: partial( dismissTip, id ),
```
FYI: `partial` is [roadmapped for removal](https://github.com/lodash/lodash/wiki/Roadmap#v500-2018) in the next major release of Lodash.

noisysocks replied:
So long old friend 😢 13dd70961

aduth commented on nux/components/dot-tip/README.md:
```diff
@@ -0,0 +1,31 @@
+
+### id
+
+An identifier that uniquely identifies the tip.
```
Should we want to establish/document naming conventions?

noisysocks replied:
👌 524f5b8db

aduth commented on nux/store/actions.js:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @return {Object} Action object.
+ */
+export function triggerGuide( tipIDs ) {
```
See bikeshed at #6741 . I think `tipIds` might be most consistent / forward-compatible here.

noisysocks replied:
👌 7857e75fa

aduth commented on nux/store/reducer.js:
```diff
@@ -0,0 +1,41 @@
+ */
+import { combineReducers } from '@wordpress/data';
+
+export function guides( state = [], action ) {
```
Would have liked to see JSDoc for each of these reducers to explain what they're tracking.

noisysocks replied:
👌 6af12cbd2

aduth commented on nux/store/selectors.js:
```diff
@@ -0,0 +1,57 @@
+ * @property {?string}  currentTipID The guide's currently showing tip.
+ * @property {?string}  nextTipID    The guide's next tip to show.
+ *
+ * @return {?NUX.GuideInfo} Information about the associated guide.
```
Where can I read more about this `NUX.GuideInfo` type ? (We should want to [`@typedef`](http://usejsdoc.org/tags-typedef.html) any of our custom types)

noisysocks replied:
It's defined above. I've moved the `@typedef` into its own block in 6af12cbd2

aduth commented on nux/store/selectors.js:
```diff
@@ -0,0 +1,57 @@
+		if ( includes( tipIDs, tipID ) ) {
+			const nonDismissedTips = difference( tipIDs, keys( state.preferences.dismissedTips ) );
+			const [ currentTipID = null, nextTipID = null ] = nonDismissedTips;
+			return { tipIDs, currentTipID, nextTipID };
```
Note: This function returns a new object every time it's called, meaning it will force React to reconcile every render if used directly in a component. This is not the case currently, but noting as a potential future performance concern (`createSelector` is used elsewhere as a memoization technique). Or we don't return an object and instead create separate selectors for each property returned here.

aduth commented on test/e2e/support/utils.js:
```diff
@@ -61,6 +61,11 @@ export async function visitAdmin( adminPath, query ) {
 export async function newPost( postType ) {
 	await visitAdmin( 'post-new.php', postType ? 'post_type=' + postType : '' );
+
+	// Disable new user tips so that their UI doesn't get in the way
```
A NUX seems like a good candidate for E2E testing, since a developer is unlikely to encounter it in their day-to-day work, and thus is more prone to accidental breakage.

noisysocks replied:
Definitely. I'll follow up on this later though since the PR is already quite large. Issue to track: https://github.com/WordPress/gutenberg/issues/7037

aduth reviewed (commented):
Very cool. Code looks generally good too. Nice one.

afercia commented:
@karmatosed @folletto well I'd agree the bar + button would be too heavy. Maybe something less invasive could help. However, I still think a tips box that opens on page load is not so different from the controversial "pointers" in WordPress (a11y considerations apart).

noisysocks committed:
Don't encourage using domReady in the NUX README

noisysocks committed:
Add wp-nux as a dependency of wp-edit-post

noisysocks committed:
Suggest that NUX tip identifiers are prefixed

noisysocks committed:
Don't use partial(), which is slated for removal

noisysocks committed:
tipID → tipId

noisysocks committed:
Improve NUX JSDoc comments

noisysocks committed:
Memoize the getAssociatedGuide() selector

noisysocks committed:
NUX: Dismiss the tip when one clicks away from it

noisysocks committed:
Position tooltips above NUX tips

noisysocks committed:
Remove aria-modal property from DotTip

Safari 11 has a weird bug when this property is used.

noisysocks committed:
Make order of the markup in DotTip match how they appear visually

noisysocks committed:
Improve a11y of DotTip labels

noisysocks committed:
Implement New User Guide

Adds a series of floating modal 'tips' which introduces a new user to
the editor. These tips can be advanced one by one or dismissed
alltogether. If dismissed, they will never show again.

noisysocks commented:
So many comments! Thanks everyone 😍 

I've responded and/or addressed all of the inline comments—thanks @aduth and @youknowriad.

In an effort to keep the size of this PR under control, I split the Popover changes out into https://github.com/WordPress/gutenberg/pull/7038.

---

> > Right now you can click outside the box, even when it's showing. This can lead to all sorts of weird things I'm sure. The box is not modal, in other words.
> I think having clicking outside to dismiss makes sense.

👌 done in 0322561.

> > I like tips. In video games tips are usually togglable in an Options menu. Should we make an option in the More menu that is enabled by default for new users, and is disabled when you're done or dismiss it?
> Yes this was in the original spec and I would like to have it included. Maybe w can make sure this gets into v1 if desired?

I really like this idea but let's do it in a follow-up PR. I want to just focus on polishing the very minimal MVP that we have for the moment. Created a follow-up issue here: https://github.com/WordPress/gutenberg/issues/7039

> I get a warning React does not recognize thenoArrowprop on a DOM element so I guess this prop gets passed to some DOM element

👌 fixed over in https://github.com/WordPress/gutenberg/pull/7038.

> the Close button tooltip has a minor z-index issue (its "arrow" is hidden by the tip box)

👌 fixed in 4890853.

> I'm not sure the "Close" button should disable the tips; it looks like a Close button, so maybe it should just close the modal. I'd consider a separate UI control to disable the tips

Thoughts, @karmatosed?

> initial focus: when I click on New Post, the initial tip appears but focus is at the root of the document; this is not so ideal for keyboard and assistive technology users. In the same way the tips are visually prominent and draw attention, they should also be the first content users perceive so initial focus should go to the modal

I'm working on this. Likely, whatever we do to address https://github.com/WordPress/gutenberg/pull/6631#discussion_r191994677 will also fix this.

> despite my initial recommendation, we should remove aria-modal because Safari 11 has a weird bug

👌 fixed in 0a45d9d.

> labels: I'd consider to improve the wording

👌 fixed in ae2a1cc.

karmatosed commented:
> I'm not sure the "Close" button should disable the tips; it looks like a Close button, so maybe it should just close the modal. I'd consider a separate UI control to disable the tips

I agree lets have this as closing the modal and the UI here:

![2018-05-31 at 12 49](https://user-images.githubusercontent.com/253067/40780684-1f3df70e-64d1-11e8-923c-bfe44d46e07e.png)

To confirm we should have the following behaviour:
- Close turns off that Tip.
- Clicking out of the Tip closes that current Tip.

This doesn't have to be in version one but it totally is the next step:
- This menu item disables Tips totally (If off we can have this say 'Enable Tips' and that turns them on - this sets the foundation for work going ahead.

For version one let's not have it have a step to start, we can iterate and test from here. I feel we need to right now get this in and work from there.

karmatosed reviewed (approved):
Let's get this into 3.0 as a first version. We have come a long way and can iterate.

noisysocks committed:
Dismiss tips instead of disabling when the X is clicked

noisysocks committed:
Fix DotTip focus issue

The first DotTip should receive focus when the page loads, and focus
should not be on the X button.

Also improve the note explaining our temporary position workaround.

youknowriad commented:
Noticed a bug if you already used Gutenberg and opened a post with several empty paragraphs

<img width="855" alt="screen shot 2018-06-01 at 10 12 21" src="https://user-images.githubusercontent.com/272444/40832954-69ef2a44-6584-11e8-8535-510d6b163092.png">

youknowriad reviewed (changes requested):
There's still a bug ⬆️ 

While I'd love to get this in 3.0, I feel we shouldn't rush it as a last minute change and merge early right after the release.

noisysocks committed:
NUX: Only show inserter tip in DefaultBlockAppender

Only display the first DotTip in the new user guide in the
DefaultBlockAppender that appears in a new post. This prevents the guide
from appearing in an existing post that contains empty blocks.

youknowriad reviewed (approved):
LGTM 👍 Let's get this early for 3.1

noisysocks commented:
Thanks all! ❤️

afercia commented:
Sorry for late reply:

> What should be focused, @afercia? Just the modal itself? Right now, Popover focuses the first focusable element in the container element which is the button.

Considering there's also the order issue:

> Close button position: visual order should always match DOM order; currently, this button is at the end of the tip content but then positioned with CSS at the top

moving the "X" button at the beginning of the DOM would solve it, as it would be the first focusable element, followed by the content. Should I open a separate issue? @noisysocks

noisysocks commented:
> moving the "X" button at the beginning of the DOM would solve it, as it would be the first focusable element, followed by the content. Should I open a separate issue? @noisysocks

No need for an issue, I can do this in a follow-up PR 🙂

noisysocks commented:
@afercia: Hm, the issue with moving the X button to the beginning of the DOM is that the tooltip is then shown when a new user opens Gutenberg for the first time:

<img width="647" alt="nux-focus" src="https://user-images.githubusercontent.com/612155/41083042-8960d08c-6a73-11e8-86d6-ad679f0ddc9c.png">

That's a little jarring. Also, I think that we _want_ the 'See next tip' button to be initially focused as it is the primary action that we want to encourage. That is, we want to encourage folks to advance through the tutorial and not dismiss the guide entirely.

Does that sound right to you? Do you have any alternative suggestions?

afercia commented:
Hm @noisysocks well "visual order should match DOM order" it's part of a specific WCAG requirement "when navigation sequences affect meaning or operation". On the other hand, there may be different orders that reflect logical relationships in the content so I guess there may be different opinions.

Reference:
Success Criterion 2.4.3 Focus Order
https://www.w3.org/TR/WCAG21/#focus-order

Understanding Success Criterion 2.4.3: Focus Order
https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html

Making the DOM order match the visual order
https://www.w3.org/WAI/WCAG21/Techniques/css/C27

The important bit is that focus order should not appear to "jump around" randomly. That said, there's a technique for modals that can be used to make screen readers announce the content automatically using `aria-describedby`:

```
<Popover
	ref={ this.popoverRef }
	className="nux-dot-tip"
	position="middle right"
	noArrow
	focusOnMount
	role="dialog"
	aria-label={ __( 'Gutenberg tips' ) }
	onClose={ onDismiss }
	onClick={ ( event ) => event.stopPropagation() }
	aria-describedby="tip-content-id-here"
>
	<p id="tip-content-id-here">{ children }</p>
```

Of course the ID should be unique. Multiple elements can be referenced too: `aria-describedby="first-element-id second-element-id"`

Aside: I've just noticed that pressing Escape navigates to the next tip, not sure why. The expected behavior is that it should close the tip.

aduth commented on nux/components/dot-tip/index.js:
```diff
@@ -0,0 +1,95 @@
+			>
+				<p>{ children }</p>
+				<p>
+					<Button isLink onClick={ onDismiss }>
```
```
Warning: validateDOMNesting(...): <button> cannot appear as a descendant of <button>.
    in button
    in p (created by DotTip)
```

noisysocks replied:
When do you see that warning, @aduth?

aduth replied:
I think it's with the default block appender `Inserter` tip, whose children are rendered within an `IconButton`:

https://github.com/WordPress/gutenberg/blob/c6cbfb041cf8c7860cbeaf1c36273d068e5e839e/editor/components/default-block-appender/index.js#L60-L64

https://github.com/WordPress/gutenberg/blob/c6cbfb041cf8c7860cbeaf1c36273d068e5e839e/editor/components/inserter/index.js#L60-L70

aduth commented on nux/components/dot-tip/index.js:
```diff
@@ -0,0 +1,95 @@
+				role="dialog"
+				aria-label={ __( 'Gutenberg tips' ) }
+				onClose={ onDismiss }
+				onClick={ ( event ) => event.stopPropagation() }
```
Why do we stop propagation.

aduth replied:
>Tips are often nested within buttons. We stop propagation so that clicking on a tip doesn't result in the button being clicked.

This reads perfectly as an inline code comment 😄 (We should be cautious about our use of `stopPropagation. This is a reasonable one, though not clear to the future maintainer why it exists)

noisysocks replied:
😀 yep, you're right: https://github.com/WordPress/gutenberg/pull/8339

afercia commented:
> Tips are often nested within buttons.

Yes and since the tips contain buttons, and they're initially within buttons, as noted in #7510 this produces a `Warning: validateDOMNesting(...): <button> cannot appear as a descendant of <button>` (at least for me)

noisysocks commented:
Very weird that that warning happens. The React component is a descendent a React `<button>`, element, sure, but the actual rendered DOM nodes are never within any `<button>` DOM nodes since we are using portals. I wonder if there's a React bug causing the warning.

afercia commented:
Not sure, maybe `validateDOMNesting` does a static analysis before the portal kicks in? Just a guess

aduth commented:
>sure, but the actual rendered DOM nodes are never within any `<button>` DOM nodes since we are using portals.

Important to note that `Popover` doesn't _force_ the use of portals. It also allows an inline rendering mode when the context is not provided (see `render`). 

It's possible this could be related to findings [described at #6799](https://github.com/WordPress/gutenberg/pull/6799/files#r206211471), where there's an initial render of Popover without the slot portaling.

Or, as @afercia mentioned, it may well be that React performs its analysis on the virtual shape, not necessarily what's rendered to the DOM.
